File: Components/Strata.Client.Core/main/typescript/Presenter/Action.ts
import {IAction} from "./IAction";
import {ICompletionStage} from "strata.foundation.core";

export
class Action<M>
    implements IAction<M>
{
    private readonly key: string;
    private readonly action: (model: M) => ICompletionStage<M>;

    constructor(key: string,action: (model: M) => ICompletionStage<M>)
    {
        this.key = key;
        this.action = action;
    }

    getKey(): string
    {
        return this.key;
    }

    apply(model: M): ICompletionStage<M>
    {
        return this.action(model);
    }

}

File: Components/Strata.Client.Core/src/main/java/strata/client/core/presenter/Action.java
//////////////////////////////////////////////////////////////////////////////
// LambdaAction.java
//////////////////////////////////////////////////////////////////////////////

package strata.client.core.presenter;

import java.util.concurrent.CompletionStage;
import java.util.function.Function;

public
class Action<M>
    implements IAction<M>
{
    private final Class<M>                       itsKey;
    private final Function<M,CompletionStage<M>> itsAction;

    public
    Action(Class<M> key,Function<M,CompletionStage<M>> action)
    {
        itsKey = key;
        itsAction = action;
    }

    @Override
    public Class<M>
    getKey()
    {
        return itsKey;
    }

    @Override
    public CompletionStage<M>
    apply(M model)
    {
        return itsAction.apply(model);
    }
}

//////////////////////////////////////////////////////////////////////////////
